LingoClash/LingoClash/PKGameView/PKGameLobby/FirebasePKGameMatchFinder.swift

```diff
@@ -0,0 +1,186 @@
+//
+//  FirebaseMultiplayerMatchFinder.swift
+//  LingoClash
+//
+//  Created by Sherwin Poh on 3/4/22.
+//
+import PromiseKit
+import FirebaseFirestore
+import FirebaseFirestoreSwift
+
+class FirebasePKGameMatchFinder: PKGameMatchFinder {
+    enum FirebaseMultiplayerMatchFinderError: Error {
+        case invalidQuerySnapshot
+        case invalidDocumentSnapshot
+        case emptyDocument
+        case questionGenerationError
+        case noCurrentBook
+    }
+    
+    private let db = Firestore.firestore()
+    private let gameType = DataManagerResources.pkGames
+    private let bookDataManager = BookManager()
+    private let questionsGenerator = QuestionsGenerator()
+    private let queueEntryDataManager = QueueEntryDataManager()
+    private let pkGameDataManager = PKGameManager()
+    private let profileDataManager = ProfileManager()
+    func findGame(playerProfile: Profile) -> Promise<PKGame> {
+        return firstly {
+            getValidQueueEntries(playerProfile: playerProfile)
+        }.then { [self] queueEntries -> Promise<PKGame> in
+            if queueEntries.count < PKGame.playerCountRequired - 1 {
+                return self.joinQueue(playerProfile: playerProfile)
+            } else {
+                return self.createGame(playerProfile: playerProfile, queueEntries: queueEntries)
+            }
+        }
+    }
+    
+    private func getValidQueueEntries(playerProfile: Profile) -> Promise<[QueueEntry]> {
+        //    1. Find all queueEntries that have:
+        //        a. Same game type
+        //        b. isWaiting is true
+        //        c. Different playerId
+        //        d. Overlap in playerScope
+        guard let currentBookId = playerProfile.currentBook?.id else {
+            return Promise.reject(reason: FirebaseMultiplayerMatchFinderError.noCurrentBook)
+        }
+        let filters: [String: Any] = [
+            "gameType": self.gameType,
+            "isWaiting": true,
+            "playerProfile.book_id": currentBookId
+        ]
+        return queueEntryDataManager.getList(field: "joinedAt", isDescending: false, filter: filters)
+    }
+    
+    private func createGame(playerProfile: Profile, queueEntries: [QueueEntry]) -> Promise<PKGame> {
+        return firstly {
+            generatePKGameData(playerProfile: playerProfile, queueEntries: queueEntries)
+        }.then { [self] pkGameData in
+            self.pkGameDataManager.createPKGame(pkGameData: pkGameData)
+        }.then { [self] newGame -> Promise<PKGame> in
+            let batch = self.db.batch()
+            for id in queueEntries.map({ $0.id }) {
```

Comment:
TODO: refactor this to be in dataManager layer for abstraction to be consistent with rest.